plugins/rts54hub/fu-rts54hub-rtd21xx-foreground.c

```diff
@@ -126,33 +126,46 @@ fu_rts54hub_rtd21xx_foreground_attach (FuDevice *device, GError **error)
 {
 	FuRts54HubDevice *parent = FU_RTS54HUB_DEVICE (fu_device_get_parent (device));
 	FuRts54hubRtd21xxForeground *self = FU_RTS54HUB_RTD21XX_FOREGROUND (device);
-	guint8 buf[] = { ISP_CMD_FW_UPDATE_RESET };
+	guint8 write_buf[ISP_PACKET_SIZE] = { 0x0 };
 	g_autoptr(FuDeviceLocker) locker = NULL;
 
 	/* open device */
 	locker = fu_device_locker_new (parent, error);
 	if (locker == NULL)
 		return FALSE;
 
-	/* exit fw mode */
-	fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);
-	fu_device_set_progress (device, 0);
-	if (!fu_rts54hub_rtd21xx_device_read_status (FU_RTS54HUB_RTD21XX_DEVICE (self),
-						     NULL, error))
-		return FALSE;
-	if (!fu_rts54hub_rtd21xx_device_i2c_write (FU_RTS54HUB_RTD21XX_DEVICE (self),
+	if (!fu_device_has_flag (device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER)) {
+		write_buf[0] = ISP_CMD_FW_UPDATE_EXIT;
+		if (!fu_rts54hub_rtd21xx_device_i2c_write (FU_RTS54HUB_RTD21XX_DEVICE (self),
 						   UC_ISP_SLAVE_ADDR,
 						   UC_FOREGROUND_OPCODE,
-						   buf, sizeof(buf),
+						   write_buf, 1,
 						   error)) {
-		g_prefix_error (error, "failed to ISP_CMD_FW_UPDATE_RESET: ");
-		return FALSE;
-	}
-
-	/* the device needs some time to restart with the new firmware before
-	* it can be queried again */
-	fu_device_sleep_with_progress (device, 60);
+			g_prefix_error (error, "failed to ISP_CMD_FW_UPDATE_RESET: ");
+			return FALSE;
+		}
+	} else {
+		/* exit fw mode */
+		fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);
+		fu_device_set_progress (device, 0);
+		if (!fu_rts54hub_rtd21xx_device_read_status (FU_RTS54HUB_RTD21XX_DEVICE (self),
+							     NULL, error))
+			return FALSE;
+		write_buf[0] = ISP_CMD_FW_UPDATE_RESET;
+		if (!fu_rts54hub_rtd21xx_device_i2c_write (FU_RTS54HUB_RTD21XX_DEVICE (self),
+							   UC_ISP_SLAVE_ADDR,
+							   UC_FOREGROUND_OPCODE,
+							   write_buf, 1,
+							   error)) {
+			g_prefix_error (error, "exit foreground-fw mode: ");
+			return FALSE;
+		}
 
+		/* the device needs some time to restart with the new firmware before
+		* it can be queried again */
+		fu_device_sleep_with_progress (device, 60);
+		fu_device_remove_flag (device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
```

Comment:
rather than hardcoding this, is it possible for setup() to either add or remove the flag depending on a query result?

Reply:
Sure, no problem. Most other plugins detect what "mode" the device is in during `fu_rts54hub_rtd21xx_foreground_setup()` -- but I guess in this case maybe you're right -- as `fu_rts54hub_rtd21xx_foreground_attach()` is being called both at setup time (because of the FuDeviceLocker) and **also** after the firmware update has been run.

Are you using the presence of `FWUPD_DEVICE_FLAG_IS_BOOTLOADER` to chose what to run during attach? Because if so, might it be a lot simpler if you had something like this in `fu_rts54hub_rtd21xx_foreground_setup()`:

	locker = fu_device_locker_new_full (device,
					    (FuDeviceLockerFunc) fu_device_detach,
					    (FuDeviceLockerFunc) fu_device_update_exit,
					    error);

and then `fu_device_update_exit()` would use `ISP_CMD_FW_UPDATE_EXIT` and the existing `fu_device_attach_for_setup()` (used after upgrade) would use just `ISP_CMD_FW_UPDATE_RESET`.

Reply:
Exactly like that, thanks.